### PR TITLE
Split string generator axioms into separate vectors

### PR DESCRIPTION
**Pull request by romainbrenguier**
This was a commit of https://github.com/diffblue/cbmc/pull/1399 which wasn't merged yet.

### CODE REVIEW

**Review by martin-cs — COMMENTED** [review]
Looks like a correct implementation but ... why?  There doesn't appear to be anything saying why this change is useful / desireable / necessary.

**Review by peterschrammel — COMMENTED** [review]
There's a fixup that needs squashing.

**Comment by peterschrammel** [comment]
A commit comment body explaining this design decision would be appreciated.

**Comment by romainbrenguier** [comment]
@martin-cs @peterschrammel I added explanations for why we do these changes in the commit message. Here is the content of the message:

> 
> Split string generator axioms into separate vectors 
>
> These axioms used to be all added in the same exprt vector by the constraint
> generator and then filtered by the solver according to their type.
> We now make the constraint generator add to appropriate vector
> according to the type of axiom (simple lemma, string constraint or "not
> contains" constraint).
> This has mainly two advantages:
>  
>   * there is no need for filtering the vector in the solver since they
>     are already prefiltered
>   * string_constraintt and string_not_contains_constraintt no longer
>     need to inherit from exprt since they are not stored in the same
>     vector.
>

**Review by martin-cs — APPROVED** [review]
#worksforme

File: src/solvers/refinement/string_refinement.cpp [hunk 1]
```diff
@@ -812,33 +812,34 @@ decision_proceduret::resultt string_refinementt::dec_solve()
   }
 
   const auto get = [this](const exprt &expr) { return this->get(expr); };
-  for(exprt axiom : generator.get_axioms())
+
+  for(string_constraintt constraint : generator.get_constraints())
```
Comment (LAJW) [inline src/solvers/refinement/string_refinement.cpp]: I'd rewrite that as `std::transform(constraints.begin(), constraints.end(), std::back_inserter(axioms.universal), []...)`

Also this for loop is missing `const` and reference (each constraint is unnecessarily copied).

File: src/solvers/refinement/string_refinement.cpp [hunk 2]
```diff
@@ -812,33 +812,34 @@ decision_proceduret::resultt string_refinementt::dec_solve()
...
+  {
+    symbol_resolve.replace_expr(constraint);
+    DATA_INVARIANT(
+      is_valid_string_constraint(error(), ns, constraint),
+      string_refinement_invariantt(
+        "string constraints satisfy their invariant"));
+    axioms.universal.push_back(constraint);
+  }
+
+  for(exprt axiom : generator.get_not_contains_constraints())
```
Comment (LAJW) [inline src/solvers/refinement/string_refinement.cpp]: Again, transform, `const` and `&` are missing.

File: src/solvers/refinement/string_refinement.cpp [hunk 3]
```diff
@@ -812,33 +812,34 @@ decision_proceduret::resultt string_refinementt::dec_solve()
...
-    }
-    else
-    {
-      add_lemma(axiom);
-    }
+    string_not_contains_constraintt nc_axiom =
+      to_string_not_contains_constraint(axiom);
+    array_typet rtype = to_array_type(nc_axiom.s0().type());
+    const typet &index_type = rtype.size().type();
+    array_typet witness_type(index_type, infinity_exprt(index_type));
```
Comment (LAJW) [inline src/solvers/refinement/string_refinement.cpp]: Explicit typing is redundant here (`arrayt x = to_array`, etc.). Use `auto` instead. Also `const`s are missing.

File: src/solvers/refinement/string_refinement.cpp [hunk 4]
```diff
@@ -812,33 +812,34 @@ decision_proceduret::resultt string_refinementt::dec_solve()
...
+    generator.witness[nc_axiom] =
+      generator.fresh_symbol("not_contains_witness", witness_type);
+    axioms.not_contains.push_back(nc_axiom);
+  }
+
+  for(exprt lemma : generator.get_lemmas())
```
Comment (LAJW) [inline src/solvers/refinement/string_refinement.cpp]: `const exprt &lemma` please.

File: unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp [hunk 5]
```diff
@@ -193,28 +193,30 @@ SCENARIO("instantiate_not_contains",
...
+        generator.get_not_contains_constraints())
+    {
+      simplify(sc, ns);
+      axioms += from_expr(ns, "", sc);
+      generator.witness[sc] = generator.fresh_symbol("w", t.witness_type());
+      nc_axioms.push_back(sc);
       axioms+="\n\n";
     }
+
+    for(exprt axiom : generator.get_lemmas())
```
Comment (LAJW) [inline unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp]: I'd use `std::accumulate` for these loops. But that's just me.

File: unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp [hunk 6]
```diff
@@ -193,28 +193,30 @@ SCENARIO("instantiate_not_contains",
...
+      axioms += "\n\n";
+    }
 
+    for(string_not_contains_constraintt sc :
+        generator.get_not_contains_constraints())
+    {
+      simplify(sc, ns);
+      axioms += from_expr(ns, "", sc);
+      generator.witness[sc] = generator.fresh_symbol("w", t.witness_type());
+      nc_axioms.push_back(sc);
```
Comment (LAJW) [inline unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp]: If you were to break this up into two loops, you could reuse a lambda for the 3 loops and turn this into `std::copy`.

**Review by LAJW — APPROVED** [review]
Looks alright as far as CBMC is concerned, I'd tweak some things.

File: unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp (hunk 6 above)
Comment (romainbrenguier) [reply unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp]: What would be the type of this lambda?  The 3 vectors have different types.

**Pull request merged**